From 64b9488535cff7ff3daf47630800265ea47d0cfb Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:38:27 -0500
Subject: [drm] i915: Fix crash when failing to parse MIPI VBT

Message-id: <1418852380-13061-17-git-send-email-rclark@redhat.com>
Patchwork-id: 102154
O-Subject: [RHEL7 drm 16/89] drm/i915: Fix crash when failing to parse MIPI VBT
Bugzilla: 1153301
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1153301
Upstream: since v3.16.4

commit 1fee660a28f88ebf9d999d6d3f2f4559a8b5eb9c

Author:     Rafael Barbalho <rafael.barbalho@intel.com>
AuthorDate: Thu Jul 24 15:16:12 2014 +0100
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Wed Dec 10 17:13:05 2014 -0500

    drm/i915: Fix crash when failing to parse MIPI VBT

    commit ed3b6679936801fa2dab47e7a6ef74e383400ed9 upstream.

    This particular nasty presented itself while trying to register the
    intelfb device (intel_fbdev.c). During the process of registering the device
    the driver will disable the crtc via i9xx_crtc_disable. These will
    also disable the panel using the generic mipi panel functions in
    dsi_mod_vbt_generic.c. The stale MIPI generic data sequence pointers would
    cause a crash within those functions. However, all of this is happening
    while console_lock is held from do_register_framebuffer inside fbcon.c. Which
    means that you got kernel log and just the device appearing to reboot/hang for
    no apparent reason.

    The fault started from the FB_EVENT_FB_REGISTERED event using the
    fb_notifier_call_chain call in fbcon.c.

    This regression has been introduced in

    commit d3b542fcfc72d7724585e3fd2c5e75351bc3df47
    Author: Shobhit Kumar <shobhit.kumar@intel.com>
    Date:   Mon Apr 14 11:00:34 2014 +0530

        drm/i915: Add parsing support for new MIPI blocks in VBT

    Cc: Shobhit Kumar <shobhit.kumar@intel.com>
    Signed-off-by: Rafael Barbalho <rafael.barbalho@intel.com>
    Reviewed-by: Shobhit Kumar <shobhit.kumar@intel.com>
    [danvet: Add regression citation.]
    Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Rob Clark <rclark@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/intel_bios.c b/drivers/gpu/drm/i915/intel_bios.c
index 827498e..621ae89 100644
--- a/drivers/gpu/drm/i915/intel_bios.c
+++ b/drivers/gpu/drm/i915/intel_bios.c
@@ -877,7 +877,7 @@ err:
 
  /* error during parsing so set all pointers to null
   * because of partial parsing */
- memset(dev_priv->vbt.dsi.sequence, 0, MIPI_SEQ_MAX);
+ memset(dev_priv->vbt.dsi.sequence, 0, sizeof(dev_priv->vbt.dsi.sequence));
 }
 
 static void parse_ddi_port(struct drm_i915_private *dev_priv, enum port port,
-- 
1.7.1